// @sap_oss/wdio-qmate-service/lib/reuse/modules/util/constants/formatter.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeFormats = exports.DateFormats = void 0;
var DateFormats;
(function (DateFormats) {
    DateFormats["MONTH_DAY_YEAR_SLASH"] = "mm/dd/yyyy";
    DateFormats["MONTH_DAY_YEAR_DASH"] = "mm-dd-yyyy";
    DateFormats["DAY_MONTH_YEAR_DOT"] = "dd.mm.yyyy";
    DateFormats["DAY_MONTH_YEAR_SLASH"] = "dd/mm/yyyy";
    DateFormats["YEAR_MONTH_DAY_PLAIN"] = "yyyymmdd";
    DateFormats["YEAR_MONTH_DAY_SLASH"] = "yyyy/mm/dd";
    DateFormats["YEAR_MONTH_DAY_DOT"] = "yyyy.mm.dd";
    DateFormats["YEAR_MONTH_DAY_DASH"] = "yyyy-mm-dd";
    DateFormats["DAY_MONTH_YEAR_TIME_DOT"] = "dd.mm.yyyy.hh.mm";
    DateFormats["MONTH_DAY_YEAR_COMMA"] = "mmm dd, yyyy";
    DateFormats["MONTH_DAY_YEAR_COMMA_SHORT"] = "mmm d, yyyy";
    DateFormats["DATETIME"] = "datetime";
    DateFormats["OBJECT"] = "object";
    DateFormats["JAPANESE_DOT"] = "g.yy.mm.dd";
    DateFormats["JAPANESE_SLASH"] = "g/yy/mm/dd";
    DateFormats["JAPANESE_DASH"] = "g-yy-mm-dd";
})(DateFormats || (exports.DateFormats = DateFormats = {}));
;
var TimeFormats;
(function (TimeFormats) {
    TimeFormats["HOUR_MINUTE_SECOND"] = "HH:mm:ss";
    TimeFormats["HOUR_MINUTE_SECOND_AM_PM"] = "h:mm:ss a";
    TimeFormats["HOUR_MINUTE_SECOND_GMT"] = "HH:mm:ss z";
    TimeFormats["HOUR_MINUTE_SECOND_AM_PM_GMT"] = "h:mm:ss a z";
    TimeFormats["HOUR_MINUTE"] = "HH:mm";
    TimeFormats["HOUR_MINUTE_AM_PM"] = "h:mm a";
    TimeFormats["HOUR"] = "HH";
    TimeFormats["HOUR_AM_PM"] = "h a";
})(TimeFormats || (exports.TimeFormats = TimeFormats = {}));
;
