// @sap_oss/wdio-qmate-service/lib/scripts/locators/qmateLocatorSrc/filters/BaseFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseFilter = void 0;
const LocatorDebug_1 = require("../utils/LocatorDebug");
class BaseFilter {
    constructor(filterFactory, rawElementProperties) {
        this.elementProperties = this.convertRawElementPropertiesToElementProperties(rawElementProperties);
        this.filterFactory = filterFactory;
        this.results = new Map();
    }
    // Public
    filter(controls) {
        if (Object.keys(this.elementProperties).length === 0 || controls.length === 0) {
            return controls;
        }
        LocatorDebug_1.LocatorDebug.beginLog(this.constructor.name, controls.length);
        const filteredControls = this.doFiltering(controls);
        LocatorDebug_1.LocatorDebug.endLog(this.constructor.name, filteredControls.length);
        return filteredControls;
    }
    checkSingle(control) {
        if (Object.keys(this.elementProperties).length === 0) {
            return true;
        }
        if (!control) {
            return false;
        }
        const controlID = control.getId();
        if (!this.results.has(controlID)) {
            this.results.set(controlID, this.doCheckSingle(control));
        }
        return this.results.get(controlID);
    }
    doFiltering(controls) {
        return controls.filter((control) => this.checkSingle(control));
    }
    // Private
    convertRawElementPropertiesToElementProperties(rawElementProperties) {
        // needed for backward compatibility
        let elementProperties = { ...rawElementProperties };
        if (typeof rawElementProperties?.mProperties === "object") {
            elementProperties = {
                ...rawElementProperties,
                ...rawElementProperties.mProperties
            };
            delete elementProperties.mProperties;
        }
        return elementProperties;
    }
}
exports.BaseFilter = BaseFilter;
