// @sap_oss/wdio-qmate-service/lib/scripts/hooks/utils/Ui5ExtensionMocker.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ui5ExtensionMocker = void 0;
class Ui5ExtensionMocker {
    static async mockRequests() {
        for (const response of this.mockResponses) {
            const mock = await browser.mock(response.requestUrlRegexp);
            mock.respond(response.payload, response.params);
        }
    }
}
exports.Ui5ExtensionMocker = Ui5ExtensionMocker;
_a = Ui5ExtensionMocker;
Ui5ExtensionMocker.payload = "File intentionally mocked by Qmate to improve performance";
Ui5ExtensionMocker.mockResponses = [
    {
        requestUrlRegexp: /.*\/xRayControls\/.*/gm,
        payload: _a.payload,
        params: {
            statusCode: 404,
            fetchResponse: false
        }
    },
    {
        requestUrlRegexp: /.*joule.*/gm,
        payload: _a.payload,
        params: {
            statusCode: 404,
            fetchResponse: false
        }
    },
    {
        requestUrlRegexp: /.*walkme.*/gm,
        payload: _a.payload,
        params: {
            statusCode: 404,
            fetchResponse: false
        }
    }
];
