// @sap_oss/wdio-qmate-service/lib/reuse/helper/verboseLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerboseLoggerFactory = exports.InactiveLogger = exports.ActiveLogger = void 0;
class ActiveLogger {
    constructor(invoker, prefix) {
        this.invoker = invoker;
        this.prefix = prefix;
    }
    log(message) {
        console.log(`${this.prefix}[${this.invoker.name}]: ${message}`);
    }
    error(e) {
        console.log(`${this.prefix}[${this.invoker.name}]`);
        throw e;
    }
}
exports.ActiveLogger = ActiveLogger;
class InactiveLogger {
    log(_message) {
        return;
    }
}
exports.InactiveLogger = InactiveLogger;
class VerboseLoggerFactory {
    constructor(namespace, moduleName) {
        this.namespace = namespace;
        this.moduleName = moduleName;
        this.isActive = process.env.QMATE_VERBOSE_MODE === "true" ? true : false;
        this.prefix = `[${this.namespace}][${this.moduleName}]`;
    }
    initLog(invoker) {
        if (!this.isActive)
            return new InactiveLogger();
        console.log(`${this.prefix}[${invoker.name}]: invoked`);
        return new ActiveLogger(invoker, this.prefix);
    }
}
exports.VerboseLoggerFactory = VerboseLoggerFactory;
