// @sap_oss/wdio-qmate-service/lib/reuse/modules/util/Util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Util = void 0;
// Components
const browser_1 = __importDefault(require("./browser"));
const console_1 = __importDefault(require("./console"));
const data_1 = __importDefault(require("./data"));
const file_1 = __importDefault(require("./file"));
const formatter_1 = __importDefault(require("./formatter"));
const function_1 = __importDefault(require("./function"));
const system_1 = __importDefault(require("./system"));
const component_1 = __importDefault(require("./component"));
const userSettings_1 = __importDefault(require("./userSettings"));
class Util {
    constructor() {
        this.browser = browser_1.default;
        this.console = console_1.default;
        this.data = data_1.default;
        this.file = file_1.default;
        this.formatter = formatter_1.default;
        this.function = function_1.default;
        this.system = system_1.default;
        this.component = component_1.default;
        this.userSettings = userSettings_1.default;
    }
}
exports.Util = Util;
exports.default = new Util();
