// @sap_oss/wdio-qmate-service/lib/reuse/modules/ui5/Ui5.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ui5 = void 0;
const assertion_1 = __importDefault(require("./assertion"));
const confirmationDialog_1 = __importDefault(require("./confirmationDialog"));
const control_1 = __importDefault(require("./control"));
const date_1 = __importDefault(require("./date"));
const element_1 = __importDefault(require("./element"));
const errorDialog_1 = __importDefault(require("./errorDialog"));
const footerBar_1 = __importDefault(require("./footerBar"));
const mockserver_1 = __importDefault(require("./mockserver"));
const navigation_1 = __importDefault(require("./navigation"));
const navigationBar_1 = __importDefault(require("./navigationBar"));
const qunit_1 = __importDefault(require("./qunit"));
const session_1 = __importDefault(require("./session"));
const table_1 = __importDefault(require("./table"));
const userInteraction_1 = __importDefault(require("./userInteraction"));
class Ui5 {
    constructor() {
        this.assertion = assertion_1.default;
        this.confirmationDialog = confirmationDialog_1.default;
        this.control = control_1.default;
        this.date = date_1.default;
        this.element = element_1.default;
        this.errorDialog = errorDialog_1.default;
        this.footerBar = footerBar_1.default;
        this.mockserver = mockserver_1.default;
        this.navigation = navigation_1.default;
        this.navigationBar = navigationBar_1.default;
        this.qunit = qunit_1.default;
        this.session = session_1.default;
        this.table = table_1.default;
        this.userInteraction = userInteraction_1.default;
    }
}
exports.Ui5 = Ui5;
exports.default = new Ui5();
