// @sap_oss/wdio-qmate-service/lib/reuse/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// functions
const runner_1 = require("./runner/runner");
// modules
const Util_1 = __importDefault(require("./modules/util/Util"));
const Common_1 = __importDefault(require("./modules/common/Common"));
const Ui5_1 = __importDefault(require("./modules/ui5/Ui5"));
const NonUi5_1 = __importDefault(require("./modules/nonUi5/NonUi5"));
const Service_1 = __importDefault(require("./modules/service/Service"));
const Mobile_1 = __importDefault(require("./modules/mobile/Mobile"));
const Flp_1 = __importDefault(require("./modules/flp/Flp"));
// data
const authenticators_json_1 = __importDefault(require("./data/authenticators.json"));
class ReuseLibrary {
    load() {
        // Runner functions
        // @ts-ignore
        global.cit = runner_1.cit;
        /**
         * @global
         * @description Global namespace for common modules.
         */
        const common = {
            assertion: Common_1.default.assertion,
            date: Common_1.default.date,
            navigation: Common_1.default.navigation,
            userInteraction: Common_1.default.userInteraction
        };
        global.common = {
            ...common,
            ...global.common
        };
        /**
         * @global
         * @description Global namespace for util modules.
         */
        const util = {
            browser: Util_1.default.browser,
            console: Util_1.default.console,
            data: Util_1.default.data,
            file: Util_1.default.file,
            formatter: Util_1.default.formatter,
            function: Util_1.default.function,
            system: Util_1.default.system,
            component: Util_1.default.component,
            userSettings: Util_1.default.userSettings
        };
        global.util = {
            ...util,
            ...global.util
        };
        /**
         * @global
         * @description Global namespace for UI5 modules.
         */
        const ui5 = {
            assertion: Ui5_1.default.assertion,
            confirmationDialog: Ui5_1.default.confirmationDialog,
            control: Ui5_1.default.control,
            date: Ui5_1.default.date,
            errorDialog: Ui5_1.default.errorDialog,
            element: Ui5_1.default.element,
            footerBar: Ui5_1.default.footerBar,
            mockserver: Ui5_1.default.mockserver,
            navigation: Ui5_1.default.navigation,
            navigationBar: Ui5_1.default.navigationBar,
            session: Ui5_1.default.session,
            table: Ui5_1.default.table,
            userInteraction: Ui5_1.default.userInteraction,
            qunit: Ui5_1.default.qunit,
            // data
            authenticators: authenticators_json_1.default
        };
        global.ui5 = {
            ...ui5,
            ...global.ui5
        };
        /**
         * @global
         * @description Global namespace for non UI5 modules.
         */
        const nonUi5 = {
            assertion: NonUi5_1.default.assertion,
            element: NonUi5_1.default.element,
            navigation: NonUi5_1.default.navigation,
            userInteraction: NonUi5_1.default.userInteraction,
            session: NonUi5_1.default.session
        };
        global.nonUi5 = {
            ...nonUi5,
            ...global.nonUi5
        };
        /**
         * @global
         * @description Global namespace for service modules.
         */
        const service = {
            odata: Service_1.default.odata,
            rest: Service_1.default.rest
        };
        global.service = {
            ...service,
            ...global.service
        };
        /**
         * @global
         * @description Global namespace for mobile modules.
         */
        const mobile = {
            element: Mobile_1.default.element,
            userInteraction: Mobile_1.default.userInteraction,
            gestures: Mobile_1.default.gestures,
            device: Mobile_1.default.device,
            android: Mobile_1.default.android,
            ios: Mobile_1.default.ios
        };
        global.mobile = {
            ...mobile,
            ...global.mobile
        };
        /**
         * @global
         * @description Global namespace for Fiori Launchpad (FLP) modules.
         */
        const flp = {
            userSettings: Flp_1.default.userSettings,
            userLocks: Flp_1.default.userLocks
        };
        global.flp = {
            ...flp,
            ...global.flp
        };
    }
}
exports.default = new ReuseLibrary();
